07_errors/alextmz/syncstore.go

```diff
@@ -0,0 +1,61 @@
+package main
+
+import (
+	"sync"
+)
+
+type SyncStore struct {
+	sync.Map
+	sync.Mutex
+}
+
+func NewSyncStore() *SyncStore {
+	a := SyncStore{}
+	return &a
+}
+
+func (m *SyncStore) CreatePuppy(p *Puppy) error {
+	if _, ok := m.Load(p.ID); !ok {
+		if p.Value < 0 {
+			return NewError(ErrValueLessThanZero)
+		}
+		m.Lock()
```

Comment:
Locking should be done before `if _, ok := m.Load(p.ID); !ok {....}`

Reply:
Also use defer for unlock.

```go
func (m *SyncStore) CreatePuppy(p *Puppy) error {
    if p.Value < 0 {
        return NewError(ErrValueLessThanZero)
    }

    m.Lock()
    defer m.Unlock()

    if _, ok := m.Load(p.ID); !ok {
        m.Store(p.ID, *p)
        return nil
    }

    return NewError(ErrIDBeingCreatedAlreadyExists)
}
```

Reply:
Thanks, in the new code this I won't need to do a Load and the Locks will be in the proper places.